{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "228350f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies ‘bitops’, ‘RCurl’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages('h2o')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2c8e36f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages('matrixStats')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "858de974",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages('ape')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "43281bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies ‘pixmap’, ‘sp’, ‘ade4’, ‘segmented’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages('seqinr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca97595b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "----------------------------------------------------------------------\n",
      "\n",
      "Your next step is to start H2O:\n",
      "    > h2o.init()\n",
      "\n",
      "For H2O package documentation, ask for help:\n",
      "    > ??h2o\n",
      "\n",
      "After starting H2O, you can use the Web UI at http://localhost:54321\n",
      "For more information visit https://docs.h2o.ai\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘h2o’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    cor, sd, var\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    &&, %*%, %in%, ||, apply, as.factor, as.numeric, colnames,\n",
      "    colnames<-, ifelse, is.character, is.factor, is.numeric, log,\n",
      "    log10, log1p, log2, round, signif, trunc\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(plyr)\n",
    "library(h2o) # https://www.h2o.ai/products/h2o/\n",
    "library(dplyr)\n",
    "library(reshape2)\n",
    "library(matrixStats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "00008e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rm(list=ls())\n",
    "#setwd(\"COVID-Reservoir-Prediciton\") # Set local working directory where files are located\n",
    "#setwd(\"~/GoogleDrive/Boulot/With Alain Zemkoho_machine learning/prediction of reservoirs/reservoir-prediction-scripts-data/merged_new.csv\")\n",
    "\n",
    "  library(plyr)\n",
    "  #library(h2o) # https://www.h2o.ai/products/h2o/\n",
    "  library(dplyr)\n",
    "  library(reshape2)\n",
    "  library(ape)\n",
    "  library(seqinr)\n",
    "  library(matrixStats)\n",
    "  `%not in%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))\n",
    "  \n",
    "  # Start h2o JVM\n",
    "  #localh20<-h2o.init(nthreads = -1)  # Start a local H2O cluster using nthreads = num available cores\n",
    "  \n",
    "  # Read data from file\n",
    "  f1<-read.csv(file=\"merged_gen_tfidf.csv\",header=T)\n",
    "  allP<-read.fasta(file =\"test.fas\",  seqtype = \"DNA\", as.string = TRUE, seqonly = F, strip.desc = T)\n",
    "  fis<-read.csv(file=\"bab_featureImportance_reservoir.csv\",header=T)\n",
    "  \n",
    "  fsars2<-read.csv(file=\"sars_babayan_features.csv\",header=T)\n",
    "  fsars2 <- rename(fsars2, Genbank.accession=SeqName)\n",
    "  fsars2 <- subset(fsars2, select = -c(X)) # remove X\n",
    "  sars2fas<-read.fasta(file =\"sars_wholegenome.fasta\",  seqtype = \"DNA\", as.string = TRUE, seqonly = F, strip.desc = T)\n",
    "  \n",
    "  # Feature definition\n",
    "  dinucs<-grep(\"[A|T|G|C|U]p[A|T|G|C|U]\",names(f1),value=T)\n",
    "  cps<-grep(\".[A|C|D|E|F|G|H|I|K|L|M|N|P|Q|R|S|T|V|W|X|Y]..[A|T|G|C|U]\",names(f1),value=T)\n",
    "  aa.codon.bias<-grep(\".Bias\",names(f1),value=T)\n",
    "  \n",
    "  # Feature selection (simplify dataset to required columns)\n",
    "  \n",
    "  # Run the script for 25 50 100 and 200  features.\n",
    "  nfeats<-25  # then tr y 100, 75, 25\n",
    "  \n",
    "  totalfeats<-length(fis$vimean)\n",
    "  f<-seq(from = totalfeats-(nfeats-1),to = totalfeats, by=1)\n",
    "  gen.feats<-as.character(fis$X[f])\n",
    "  \n",
    "  f1<-f1[,c(\"Genbank.accession\",\"Reservoir\", gen.feats)]\n",
    "  \n",
    "  # Remove orphans\n",
    "  f2<-subset(f1,f1$Reservoir!=\"Orphan\")\n",
    "  f<-droplevels(f2)\n",
    "  \n",
    "    # Group selection based on thresholds\n",
    "  t<-15 # threshold for minimum sample size of groups\n",
    "  s<-.85 # proportion in the training set\n",
    "  \n",
    "  host.counts<-table(f$Reservoir)\n",
    "  \n",
    "  min.t<-host.counts[host.counts>=t] # minimum number of viruses per host group\n",
    "  f_st3<-f[f$Reservoir %in% c(names(min.t)),]\n",
    "  f_st3<-droplevels(f_st3)\n",
    "  f_st3$SeqName2<-do.call(rbind,strsplit(as.character(f_st3$Genbank.accession),\"[.]\"))[,1]\n",
    "  \n",
    "  # Rare hosts\n",
    "  rare<-f[!f$Reservoir %in% c(names(min.t)),]\n",
    "  rare<-droplevels(rare)\n",
    "  rare$SeqName2<-do.call(rbind,strsplit(as.character(rare$Genbank.accession),\"[.]\"))[,1]\n",
    "  \n",
    "  # Number and names of host taxa\n",
    "  ntax<-length(unique(f_st3$Reservoir))\n",
    "  bp<-as.character(sort(unique(f_st3$Reservoir)))\n",
    "  \n",
    "  # sample split of training/test to get counts in each\n",
    "  \n",
    "  set.seed(78910)\n",
    "  \n",
    "  #trains<-f_st3 %>% group_by(Reservoir) %>%\n",
    "  # filter(Genbank.accession %in% sample(unique(Genbank.accession), ceiling(s*length(unique(Genbank.accession)))))\n",
    "  #testval<-subset(f_st3,!(f_st3$Genbank.accession %in% trains$Genbank.accession)) # ref numbers absent from training set\n",
    "  \n",
    "  #optims<-testval %>% group_by(Reservoir) %>%\n",
    "  #  filter(Genbank.accession %in% sample(unique(Genbank.accession), floor(.5*length(unique(Genbank.accession)))))\n",
    "  #tests<-subset(testval,!(testval$Genbank.accession %in% optims$Genbank.accession)) # ref numbers in testval set absent from test set\n",
    "  #ntest<-dim(tests)[1]\n",
    "  \n",
    "  # write orphan sequences\n",
    "  # orp<-allP[c(which(names(allP) %in% orphans$Genbank.accession))]\n",
    "  # write.fasta(orp,names(orp),file.out=\"orphanDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "  \n",
    "  # write rare host sequences\n",
    "  #rar<-allP[c(which(names(allP) %in% rare$Genbank.accession))]\n",
    "  #write.fasta(rar,names(rar),file.out=\"rareDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "  \n",
    "  # Remove unneeded files \n",
    "  #     rm(f,f1,f2,fis,rar,orp)\n",
    "  # Train many models\n",
    "  set.seed(78910)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "013e9acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "system('cd babayan',intern=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bd493beb",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in setwd(\"babayan\"): cannot change working directory\n",
     "output_type": "error",
     "traceback": [
      "Error in setwd(\"babayan\"): cannot change working directory\nTraceback:\n",
      "1. setwd(\"babayan\")"
     ]
    }
   ],
   "source": [
    "setwd(\"babayan\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ddce360c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in system(\"makeblastdb -in trainDB.fasta -dbtype nucl -parse_seqids -out allTrainingDB\", :\n",
      "“error in running command”\n",
      "Warning message in system(\"blastn -db allTrainingDB -query trainDB.fasta -out trainOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\", :\n",
      "“error in running command”\n",
      "Warning message in file(file, \"rt\"):\n",
      "“cannot open file 'trainOut.out': No such file or directory”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(file, \"rt\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(file, \"rt\"): cannot open the connection\nTraceback:\n",
      "1. read.csv(file = \"trainOut.out\", col.names = c(\"query acc.\", \"subject acc.\", \n .     \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \n .     \"q. start\", \"q. end\", \" s. start\", \" s. end\", \" evalue\", \n .     \" bit score\"), header = F)",
      "2. read.table(file = file, header = header, sep = sep, quote = quote, \n .     dec = dec, fill = fill, comment.char = comment.char, ...)",
      "3. file(file, \"rt\")"
     ]
    }
   ],
   "source": [
    "  nloops<-2\n",
    "  nfeatures<-length(gen.feats)+ntax\n",
    "  vimps<-matrix(nrow=nfeatures,ncol=nloops)\n",
    "  for (i in 1:nloops) {\n",
    "       # Stratified random sampling\n",
    "    trains<-f_st3 %>% group_by(Reservoir) %>%\n",
    "      filter(Genbank.accession %in% sample(unique(Genbank.accession), ceiling(s*length(unique(Genbank.accession)))))\n",
    "    tests<-subset(f_st3,!(f_st3$Genbank.accession %in% trains$Genbank.accession)) # ref numbers absent from training set #change testval to tests\n",
    "\n",
    "      trains<-droplevels(trains)\n",
    "      fsars2<-droplevels(fsars2)\n",
    "      tests<-droplevels(tests)\n",
    "      #optims<-droplevels(optims)\n",
    "      #   test.record[,i]<-as.character(tests$Genbank.accession)\n",
    "  \n",
    "      # Select and write sequences to local directory\n",
    "      trainSeqs<-allP[c(which(names(allP) %in% trains$Genbank.accession))] # pick sequences in the training set\n",
    "      testSeqs<-allP[c(which(names(allP) %in% tests$Genbank.accession))] # pick sequences in the validation set\n",
    "      sars2Seqs<-sars2fas[c(which(names(sars2fas) %in% fsars2$Genbank.accession))]\n",
    "      \n",
    "      #optSeqs<-allP[c(which(names(allP) %in% optims$Genbank.accession))] # pick sequences in the optimization set\n",
    "      write.fasta(testSeqs, names(testSeqs), file.out=\"testDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "      write.fasta(trainSeqs, names(trainSeqs), file.out=\"trainDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "      #write.fasta(optSeqs, names(optSeqs), file.out=\"optDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "      write.fasta(sars2Seqs, names(sars2Seqs), file.out=\"sars2DB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "  \n",
    "      # BLAST\n",
    "      system(\"makeblastdb -in trainDB.fasta -dbtype nucl -parse_seqids -out allTrainingDB\",intern=F)\n",
    "  \n",
    "      # Blast test against training\n",
    "      system(\"blastn -db allTrainingDB -query testDB.fasta -out testOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=FALSE)\n",
    "  \n",
    "      # Blast validation against training\n",
    "     # system(\"blastn -db allTrainingDB -query optDB.fasta -out optOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=FALSE)\n",
    "  \n",
    "      system(\"blastn -db allTrainingDB -query sars2DB.fasta -out sars2Out.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=FALSE)\n",
    "      \n",
    "      # Blast training against the training set (take top 5 hits)\n",
    "      system(\"blastn -db allTrainingDB -query trainDB.fasta -out trainOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=TRUE)\n",
    "  \n",
    "      # Read in Blast output for training set\n",
    "      allBlast<-read.csv(file=\"trainOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "  \n",
    "      # Summarize blast hits\n",
    "      nvir<-length(unique(allBlast$query.acc.))\n",
    "      virnames<-unique(allBlast$query.acc.)\n",
    "      ecutoff<-1E-3\n",
    "      j=1\n",
    "      d<-subset(allBlast,allBlast$query.acc.==virnames[j])\n",
    "      d2<-subset(d,d$X..identity<100)\n",
    "      d2<-subset(d2,d2$X.evalue<ecutoff)\n",
    "      # Assign equal probability across all hosts if there is no good blast hit\n",
    "      for (z in 1:1){\n",
    "          if (nrow(d2)==0){\n",
    "              blast.uc<-rep(1/ntax,ntax)\n",
    "              blast.uc<-data.frame(t(blast.uc))\n",
    "              colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc<-cbind(id,blast.uc)}\n",
    "          else {\n",
    "              dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "              dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "              hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "              hosts[is.na(hosts)]<-0\n",
    "              hosts<-t(data.frame(hosts))\n",
    "              hosts<-data.frame(hosts)\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc<-cbind(id,hosts)}}\n",
    "  \n",
    "      for (j in 2:nvir){\n",
    "          d<-subset(allBlast,allBlast$query.acc.==virnames[j])\n",
    "          d2<-subset(d,d$X..identity<100)\n",
    "          d2<-subset(d2,d2$X.evalue<ecutoff)\n",
    "          if (nrow(d2)==0){\n",
    "              blast.uc.s<-rep(1/ntax,ntax)\n",
    "              blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "              colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc.s<-cbind(id,blast.uc.s)}\n",
    "          else {\n",
    "              dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "              dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "              hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "              hosts[is.na(hosts)]<-0\n",
    "              hosts<-t(data.frame(hosts))\n",
    "              hosts<-data.frame(hosts)\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc.s<-cbind(id,hosts)\n",
    "              }\n",
    "          blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "          blast.uc[is.na(blast.uc)] <- 0\n",
    "          }\n",
    "  \n",
    "      f1_train<-merge(trains,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "      set<-c(\"Reservoir\",\"Genbank.accession\",gen.feats,bp)\n",
    "      myData=data.frame(matrix(nrow=1, ncol = length(set)))\n",
    "      f1_train=bind_rows(myData,f1_train)\n",
    "      f1_train<-f1_train[,c(set)] # this is the full training dataset with genomic features and blast probabilities\n",
    "      f1_train=f1_train[-1,]\n",
    "      f1_train[is.na(f1_train)]=0\n",
    "  \n",
    "      # Summarize blast hits from test set\n",
    "      testBlast<-read.csv(file=\"testOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "      nvir<-length(unique(testBlast$query.acc.))\n",
    "      virnames<-unique(testBlast$query.acc.)\n",
    "      ecutoff<-1E-3\n",
    "      j=1\n",
    "      d<-subset(testBlast,testBlast$query.acc.==virnames[j])\n",
    "      d2<-subset(d,d$X.evalue<ecutoff)\n",
    "  \n",
    "      for (z in 1:1){\n",
    "          if (nrow(d2)==0){\n",
    "              blast.uc<-rep(1/ntax,ntax)\n",
    "              blast.uc<-data.frame(t(blast.uc))\n",
    "              colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc<-cbind(id,blast.uc)}\n",
    "          else {\n",
    "              dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "              dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "              hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "              hosts[is.na(hosts)]<-0\n",
    "              hosts<-t(data.frame(hosts))\n",
    "              hosts<-data.frame(hosts)\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc<-cbind(id,hosts)}}\n",
    "  \n",
    "      for (j in 2:nvir){\n",
    "          d<-subset(testBlast,testBlast$query.acc.==virnames[j])\n",
    "          d2<-subset(d,d$X.evalue<ecutoff)\n",
    "          if (nrow(d2)==0){\n",
    "              blast.uc.s<-rep(1/ntax,ntax)\n",
    "              blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "              colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "              id<-as.character(virnames[j])\n",
    "              blast.uc.s<-cbind(id,blast.uc.s) }\n",
    "          else {\n",
    "              dhost<-merge(d,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "              dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "              hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "              hosts[is.na(hosts)]<-0\n",
    "              hosts<-t(data.frame(hosts))\n",
    "              hosts<-data.frame(hosts)\n",
    "              id<-as.character(d$query.acc.[1])\n",
    "              blast.uc.s<-cbind(id,hosts)}\n",
    "          blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "          blast.uc[is.na(blast.uc)] <- 0\n",
    "          }\n",
    "  \n",
    "      f1_test<-merge(tests,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "      testID<-f1_test$Virus.name\n",
    "      # adding in the values for reservoir groups with no relative support\n",
    "      set<-c(\"Reservoir\",\"Genbank.accession\",gen.feats,bp)\n",
    "      myData=data.frame(matrix(nrow=1, ncol = length(set)))\n",
    "      colnames(myData) = set\n",
    "      f1_test=bind_rows(myData,f1_test)\n",
    "      f1_test<-f1_test[,c(set)]\n",
    "      f1_test=f1_test[-1,]\n",
    "      f1_test[is.na(f1_test)]=0\n",
    "      \n",
    "      #cleaning the output for n=1 in testing set\n",
    "      if((nrow(tests)==1)&(nrow(f1_test)==3)) {\n",
    "      f1_test=f1_test[2,]\n",
    "      }\n",
    "      \n",
    "      ' Summarize blast hits from optimization set\n",
    "      optBlast<-read.csv(file=\"optOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "      nvir<-length(unique(optBlast$query.acc.))\n",
    "      virnames<-unique(optBlast$query.acc.)\n",
    "      ecutoff<-1E-3\n",
    "      j=1\n",
    "      d<-subset(optBlast,optBlast$query.acc.==virnames[j])\n",
    "      d2<-subset(d,d$X.evalue<ecutoff)\n",
    "  \n",
    "      for (z in 1:1){\n",
    "        if (nrow(d2)==0){\n",
    "          blast.uc<-rep(1/ntax,ntax)\n",
    "          blast.uc<-data.frame(t(blast.uc))\n",
    "          colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc<-cbind(id,blast.uc)}\n",
    "        else {\n",
    "          dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "          dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "          hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "          hosts[is.na(hosts)]<-0\n",
    "          hosts<-t(data.frame(hosts))\n",
    "          hosts<-data.frame(hosts)\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc<-cbind(id,hosts)}}\n",
    "  \n",
    "      for (j in 2:nvir){\n",
    "        d<-subset(optBlast,optBlast$query.acc.==virnames[j])\n",
    "        d2<-subset(d,d$X.evalue<ecutoff)\n",
    "        if (nrow(d2)==0){\n",
    "          blast.uc.s<-rep(1/ntax,ntax)\n",
    "          blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "          colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc.s<-cbind(id,blast.uc.s) }\n",
    "        else {\n",
    "          dhost<-merge(d,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "          dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "          hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "          hosts[is.na(hosts)]<-0\n",
    "          hosts<-t(data.frame(hosts))\n",
    "          hosts<-data.frame(hosts)\n",
    "          id<-as.character(d$query.acc.[1])\n",
    "          blast.uc.s<-cbind(id,hosts)}\n",
    "        blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "        blast.uc[is.na(blast.uc)] <- 0\n",
    "        }\n",
    "  \n",
    "      f1_opt<-merge(optims,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "      optID<-f1_opt$Virus.name\n",
    "      set<-c(\"Reservoir\",gen.feats,bp)\n",
    "      f1_opt<-f1_opt[,c(set)]'\n",
    "  \n",
    "      # save train, test, opt sets\n",
    "      \n",
    "      sars2Blast<-read.csv(file=\"sars2Out.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "      nvir<-length(unique(sars2Blast$query.acc.))\n",
    "      virnames<-unique(sars2Blast$query.acc.)\n",
    "      ecutoff<-1E-6\n",
    "      j=1\n",
    "      d<-subset(sars2Blast,sars2Blast$query.acc.==virnames[j])\n",
    "      d2<-subset(d,d$X.evalue<ecutoff)\n",
    "      \n",
    "      for (z in 1:1){\n",
    "        if (nrow(d2)==0){\n",
    "          blast.uc<-rep(1/ntax,ntax)\n",
    "          blast.uc<-data.frame(t(blast.uc))\n",
    "          colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc<-cbind(id,blast.uc)}\n",
    "        else {\n",
    "          dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "          dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "          hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "          hosts[is.na(hosts)]<-0\n",
    "          hosts<-t(data.frame(hosts))\n",
    "          hosts<-data.frame(hosts)\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc<-cbind(id,hosts)}}\n",
    "      \n",
    "      for (j in 2:nvir){\n",
    "        d<-subset(sars2Blast,sars2Blast$query.acc.==virnames[j])\n",
    "        d2<-subset(d,d$X.evalue<ecutoff)\n",
    "        if (nrow(d2)==0){\n",
    "          blast.uc.s<-rep(1/ntax,ntax)\n",
    "          blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "          colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "          id<-as.character(virnames[j])\n",
    "          blast.uc.s<-cbind(id,blast.uc.s) }\n",
    "        else {\n",
    "          dhost<-merge(d,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "          dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "          hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "          hosts[is.na(hosts)]<-0\n",
    "          hosts<-t(data.frame(hosts))\n",
    "          hosts<-data.frame(hosts)\n",
    "          id<-as.character(d$query.acc.[1])\n",
    "          blast.uc.s<-cbind(id,hosts)}\n",
    "        blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "        blast.uc[is.na(blast.uc)] <- 0\n",
    "      }\n",
    "      \n",
    "      f1_sars2<-merge(fsars2,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "      sars2ID<-f1_sars2$Virus.name\n",
    "      set<-c(\"Reservoir\",\"Genbank.accession\",gen.feats,bp)\n",
    "      myData=data.frame(matrix(nrow=1, ncol = length(set)))\n",
    "      colnames(myData) = set\n",
    "      f1_sars2=bind_rows(myData,f1_sars2)\n",
    "      f1_sars2<-f1_sars2[,c(set)]\n",
    "      f1_sars2=f1_sars2[-1,]\n",
    "      f1_sars2[is.na(f1_sars2)]=0\n",
    "   \n",
    "      write.csv(f1_train,file=paste(\"results/training_set\",i,\".csv\",sep=\"_\"))\n",
    "      write.csv(f1_sars2,file=paste(\"results/sars_set\",i,\".csv\",sep=\"_\"))\n",
    "      write.csv(f1_test,file=paste(\"results/test_set\",i,\".csv\",sep=\"_\"))\n",
    "   \n",
    "  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a560c0fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e06767d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61928281",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(plyr)\n",
    "library(h2o) # https://www.h2o.ai/products/h2o/\n",
    "library(dplyr)\n",
    "library(reshape2)\n",
    "library(ape)\n",
    "library(seqinr)\n",
    "library(matrixStats)\n",
    "`%not in%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))\n",
    "\n",
    "# Start h2o JVM\n",
    "localh20<-h2o.init(nthreads = -1)  # Start a local H2O cluster using nthreads = num available cores\n",
    "\n",
    "# Read data from file\n",
    "f1<-read.csv(file=\"merged_gen_tfidf.csv\",header=T)\n",
    "allP<-read.fasta(file =\"test.fas\",  seqtype = \"DNA\", as.string = TRUE, seqonly = F, strip.desc = T)\n",
    "fis<-read.csv(file=\"featureImportance_reservoir-1.csv\",header=T)\n",
    "\n",
    "# Feature definition\n",
    "dinucs<-grep(\"[A|T|G|C|U]p[A|T|G|C|U]\",names(f1),value=T)\n",
    "cps<-grep(\".[A|C|D|E|F|G|H|I|K|L|M|N|P|Q|R|S|T|V|W|X|Y]..[A|T|G|C|U]\",names(f1),value=T)\n",
    "aa.codon.bias<-grep(\".Bias\",names(f1),value=T)\n",
    "\n",
    "# Feature selection (simplify dataset to required columns)\n",
    "\n",
    "# Run the script for 25 50 100 and 200  features.\n",
    "nfeats<-25\n",
    "\n",
    "totalfeats<-length(fis$vimean)\n",
    "f<-seq(from = totalfeats-(nfeats-1),to = totalfeats, by=1)\n",
    "gen.feats<-as.character(fis$X[f])\n",
    "\n",
    "f1<-f1[,c(\"Genbank.accession\",\"Reservoir\",\"Viral.group\",\"Vector.borne\",\"Vector\", gen.feats)]\n",
    "\n",
    "# Remove orphans\n",
    "f2<-subset(f1,f1$Reservoir!=\"Orphan\")\n",
    "f<-droplevels(f2)\n",
    "\n",
    "\n",
    "# Group selection based on thresholds\n",
    "t<-15 # threshold for minimum sample size of groups\n",
    "s<-.85 # proportion in the training set\n",
    "\n",
    "host.counts<-table(f$Reservoir)\n",
    "\n",
    "min.t<-host.counts[host.counts>=t] # minimum number of viruses per host group\n",
    "f_st3<-f[f$Reservoir %in% c(names(min.t)),]\n",
    "f_st3<-droplevels(f_st3)\n",
    "f_st3$SeqName2<-do.call(rbind,strsplit(as.character(f_st3$Genbank.accession),\"[.]\"))[,1]\n",
    "\n",
    "# Rare hosts\n",
    "rare<-f[!f$Reservoir %in% c(names(min.t)),]\n",
    "rare<-droplevels(rare)\n",
    "rare$SeqName2<-do.call(rbind,strsplit(as.character(rare$Genbank.accession),\"[.]\"))[,1]\n",
    "\n",
    "# Number and names of host taxa\n",
    "ntax<-length(unique(f_st3$Reservoir))\n",
    "bp<-as.character(sort(unique(f_st3$Reservoir)))\n",
    "\n",
    "# sample split of training/test to get counts in each\n",
    "\n",
    "set.seed(78910)\n",
    "\n",
    "#trains<-f_st3 %>% group_by(Reservoir) %>%\n",
    "# filter(Genbank.accession %in% sample(unique(Genbank.accession), ceiling(s*length(unique(Genbank.accession)))))\n",
    "#testval<-subset(f_st3,!(f_st3$Genbank.accession %in% trains$Genbank.accession)) # ref numbers absent from training set\n",
    "\n",
    "#optims<-testval %>% group_by(Reservoir) %>%\n",
    "#  filter(Genbank.accession %in% sample(unique(Genbank.accession), floor(.5*length(unique(Genbank.accession)))))\n",
    "#tests<-subset(testval,!(testval$Genbank.accession %in% optims$Genbank.accession)) # ref numbers in testval set absent from test set\n",
    "#ntest<-dim(tests)[1]\n",
    "\n",
    "# write orphan sequences\n",
    "# orp<-allP[c(which(names(allP) %in% orphans$Genbank.accession))]\n",
    "# write.fasta(orp,names(orp),file.out=\"orphanDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "\n",
    "# write rare host sequences\n",
    "#rar<-allP[c(which(names(allP) %in% rare$Genbank.accession))]\n",
    "#write.fasta(rar,names(rar),file.out=\"rareDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "\n",
    "# Remove unneeded files \n",
    "#     rm(f,f1,f2,fis,rar,orp)\n",
    "# Train many models\n",
    "set.seed(78910)\n",
    "#change it back to 100 iterations\n",
    "nloops<-2\n",
    "#lr<-c()\n",
    "#md<-c()\n",
    "#sr<-c()\n",
    "#csr<-c()\n",
    "#nt<-c()\n",
    "# mr<-c()\n",
    "# accuracy.st3<-c()\n",
    "\n",
    "# pc.accuracy<-matrix(nrow=nloops,ncol=ntax)\n",
    "# test.record<-matrix(nrow=ntest,ncol=nloops)\n",
    "nfeatures<-length(gen.feats)+ntax\n",
    "vimps<-matrix(nrow=nfeatures,ncol=nloops)\n",
    "\n",
    "for (i in 1:nloops) {\n",
    "    # Stratified random sampling\n",
    "    trains<-f_st3 %>% group_by(Reservoir) %>%\n",
    "      filter(Genbank.accession %in% sample(unique(Genbank.accession), ceiling(s*length(unique(Genbank.accession)))))\n",
    "    tests<-subset(f_st3,!(f_st3$Genbank.accession %in% trains$Genbank.accession)) # ref numbers absent from training set #change testval to tests\n",
    "\n",
    "    #optims<-tests %>% group_by(Reservoir) %>%\n",
    "     # filter(Genbank.accession %in% sample(unique(Genbank.accession), floor(.5*length(unique(Genbank.accession)))))\n",
    "    \n",
    "   # tests<-subset(testval,!(testval$Genbank.accession %in% optims$Genbank.accession)) # ref numbers in testval set absent from test set\n",
    "\n",
    "    trains<-droplevels(trains)\n",
    "    tests<-droplevels(tests)\n",
    "    #optims<-droplevels(optims)\n",
    "    #   test.record[,i]<-as.character(tests$Genbank.accession)\n",
    "\n",
    "    # Select and write sequences to local directory\n",
    "    trainSeqs<-allP[c(which(names(allP) %in% trains$Genbank.accession))] # pick sequences in the training set\n",
    "    testSeqs<-allP[c(which(names(allP) %in% tests$Genbank.accession))] # pick sequences in the validation set\n",
    "    #optSeqs<-allP[c(which(names(allP) %in% optims$Genbank.accession))] # pick sequences in the optimization set\n",
    "    write.fasta(testSeqs, names(testSeqs), file.out=\"testDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "    write.fasta(trainSeqs, names(trainSeqs), file.out=\"trainDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "    #write.fasta(optSeqs, names(optSeqs), file.out=\"optDB.fasta\", open = \"w\", nbchar = 100, as.string = T)\n",
    "\n",
    "    # BLAST\n",
    "    system(\"makeblastdb -in trainDB.fasta -dbtype nucl -parse_seqids -out allTrainingDB\",intern=F)\n",
    "\n",
    "    # Blast test against training\n",
    "    system(\"blastn -db allTrainingDB -query testDB.fasta -out testOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=FALSE)\n",
    "\n",
    "    # Blast validation against training\n",
    "   # system(\"blastn -db allTrainingDB -query optDB.fasta -out optOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=FALSE)\n",
    "\n",
    "    # Blast training against the training set (take top 5 hits)\n",
    "    system(\"blastn -db allTrainingDB -query trainDB.fasta -out trainOut.out -num_threads 4 -outfmt 10 -max_target_seqs=5 -max_hsps 1 -reward 2 -task blastn -evalue 10 -word_size 8 -gapopen 2 -gapextend 2\",inter=F,wait=TRUE)\n",
    "\n",
    "    # Read in Blast output for training set\n",
    "    allBlast<-read.csv(file=\"trainOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "\n",
    "    # Summarize blast hits\n",
    "    nvir<-length(unique(allBlast$query.acc.))\n",
    "    virnames<-unique(allBlast$query.acc.)\n",
    "    ecutoff<-1E-3\n",
    "    j=1\n",
    "    d<-subset(allBlast,allBlast$query.acc.==virnames[j])\n",
    "    d2<-subset(d,d$X..identity<100)\n",
    "    d2<-subset(d2,d2$X.evalue<ecutoff)\n",
    "    # Assign equal probability across all hosts if there is no good blast hit\n",
    "    for (z in 1:1){\n",
    "        if (nrow(d2)==0){\n",
    "            blast.uc<-rep(1/ntax,ntax)\n",
    "            blast.uc<-data.frame(t(blast.uc))\n",
    "            colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc<-cbind(id,blast.uc)}\n",
    "        else {\n",
    "            dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "            dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "            hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "            hosts[is.na(hosts)]<-0\n",
    "            hosts<-t(data.frame(hosts))\n",
    "            hosts<-data.frame(hosts)\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc<-cbind(id,hosts)}}\n",
    "\n",
    "    for (j in 2:nvir){\n",
    "        d<-subset(allBlast,allBlast$query.acc.==virnames[j])\n",
    "        d2<-subset(d,d$X..identity<100)\n",
    "        d2<-subset(d2,d2$X.evalue<ecutoff)\n",
    "        if (nrow(d2)==0){\n",
    "            blast.uc.s<-rep(1/ntax,ntax)\n",
    "            blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "            colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc.s<-cbind(id,blast.uc.s)}\n",
    "        else {\n",
    "            dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "            dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "            hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "            hosts[is.na(hosts)]<-0\n",
    "            hosts<-t(data.frame(hosts))\n",
    "            hosts<-data.frame(hosts)\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc.s<-cbind(id,hosts)\n",
    "            }\n",
    "        blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "        blast.uc[is.na(blast.uc)] <- 0\n",
    "        }\n",
    "\n",
    "    f1_train<-merge(trains,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "    set<-c(\"Reservoir\",gen.feats,bp)\n",
    "    f1_train<-f1_train[,c(set)] # this is the full training dataset with genomic features and blast probabilities\n",
    "\n",
    "    # Summarize blast hits from test set\n",
    "    testBlast<-read.csv(file=\"testOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "    nvir<-length(unique(testBlast$query.acc.))\n",
    "    virnames<-unique(testBlast$query.acc.)\n",
    "    ecutoff<-1E-3\n",
    "    j=1\n",
    "    d<-subset(testBlast,testBlast$query.acc.==virnames[j])\n",
    "    d2<-subset(d,d$X.evalue<ecutoff)\n",
    "\n",
    "    for (z in 1:1){\n",
    "        if (nrow(d2)==0){\n",
    "            blast.uc<-rep(1/ntax,ntax)\n",
    "            blast.uc<-data.frame(t(blast.uc))\n",
    "            colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc<-cbind(id,blast.uc)}\n",
    "        else {\n",
    "            dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "            dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "            hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "            hosts[is.na(hosts)]<-0\n",
    "            hosts<-t(data.frame(hosts))\n",
    "            hosts<-data.frame(hosts)\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc<-cbind(id,hosts)}}\n",
    "\n",
    "    for (j in 2:nvir){\n",
    "        d<-subset(testBlast,testBlast$query.acc.==virnames[j])\n",
    "        d2<-subset(d,d$X.evalue<ecutoff)\n",
    "        if (nrow(d2)==0){\n",
    "            blast.uc.s<-rep(1/ntax,ntax)\n",
    "            blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "            colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "            id<-as.character(virnames[j])\n",
    "            blast.uc.s<-cbind(id,blast.uc.s) }\n",
    "        else {\n",
    "            dhost<-merge(d,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "            dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "            hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "            hosts[is.na(hosts)]<-0\n",
    "            hosts<-t(data.frame(hosts))\n",
    "            hosts<-data.frame(hosts)\n",
    "            id<-as.character(d$query.acc.[1])\n",
    "            blast.uc.s<-cbind(id,hosts)}\n",
    "        blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "        blast.uc[is.na(blast.uc)] <- 0\n",
    "        }\n",
    "\n",
    "    f1_test<-merge(tests,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "    testID<-f1_test$Virus.name\n",
    "    set<-c(\"Reservoir\",gen.feats,bp)\n",
    "    f1_test<-f1_test[,c(set)]\n",
    "\n",
    "    ' Summarize blast hits from optimization set\n",
    "    optBlast<-read.csv(file=\"optOut.out\",col.names = c(\"query acc.\", \"subject acc.\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\",\" s. start\",\" s. end\",\" evalue\",\" bit score\"),header=F)\n",
    "    nvir<-length(unique(optBlast$query.acc.))\n",
    "    virnames<-unique(optBlast$query.acc.)\n",
    "    ecutoff<-1E-3\n",
    "    j=1\n",
    "    d<-subset(optBlast,optBlast$query.acc.==virnames[j])\n",
    "    d2<-subset(d,d$X.evalue<ecutoff)\n",
    "    for (z in 1:1){\n",
    "      if (nrow(d2)==0){\n",
    "        blast.uc<-rep(1/ntax,ntax)\n",
    "        blast.uc<-data.frame(t(blast.uc))\n",
    "        colnames(blast.uc)<-sort(unique(trains$Reservoir))\n",
    "        id<-as.character(virnames[j])\n",
    "        blast.uc<-cbind(id,blast.uc)}\n",
    "      else {\n",
    "        dhost<-merge(d2,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "        dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "        hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=F)\n",
    "        hosts[is.na(hosts)]<-0\n",
    "        hosts<-t(data.frame(hosts))\n",
    "        hosts<-data.frame(hosts)\n",
    "        id<-as.character(virnames[j])\n",
    "        blast.uc<-cbind(id,hosts)}}\n",
    "    for (j in 2:nvir){\n",
    "      d<-subset(optBlast,optBlast$query.acc.==virnames[j])\n",
    "      d2<-subset(d,d$X.evalue<ecutoff)\n",
    "      if (nrow(d2)==0){\n",
    "        blast.uc.s<-rep(1/ntax,ntax)\n",
    "        blast.uc.s<-data.frame(t(blast.uc.s))\n",
    "        colnames(blast.uc.s)<-sort(unique(trains$Reservoir))\n",
    "        id<-as.character(virnames[j])\n",
    "        blast.uc.s<-cbind(id,blast.uc.s) }\n",
    "      else {\n",
    "        dhost<-merge(d,trains,by.x=\"subject.acc.\",by.y=\"Genbank.accession\",all.x = T,all.y = F)\n",
    "        dhost$rel.support<-dhost$X..identity/sum(dhost$X..identity)\n",
    "        hosts<-tapply(dhost$rel.support,dhost$Reservoir,sum,na.rm=T)\n",
    "        hosts[is.na(hosts)]<-0\n",
    "        hosts<-t(data.frame(hosts))\n",
    "        hosts<-data.frame(hosts)\n",
    "        id<-as.character(d$query.acc.[1])\n",
    "        blast.uc.s<-cbind(id,hosts)}\n",
    "      blast.uc<-rbind.fill(blast.uc,blast.uc.s)\n",
    "      blast.uc[is.na(blast.uc)] <- 0\n",
    "      }\n",
    "    f1_opt<-merge(optims,blast.uc,by.x=\"Genbank.accession\",by.y=\"id\",all.x=F,all.y=T)\n",
    "    optID<-f1_opt$Virus.name\n",
    "    set<-c(\"Reservoir\",gen.feats,bp)\n",
    "    f1_opt<-f1_opt[,c(set)]'\n",
    "\n",
    "    # save train, test, opt sets\n",
    " \n",
    "    write.csv(f1_train,file=paste(\"training_set\",i,\".csv\",sep=\"_\"))\n",
    "    #write.csv(f1_opt,file=paste(\"opt_set\",i,\".csv\",sep=\"_\"))\n",
    "    write.csv(f1_test,file=paste(\"test_set\",i,\".csv\",sep=\"_\"))\n",
    " \n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
